### DISCUSSION TITLE
Is my WebGL style optimised?

### DISCUSSION THREAD

**Discussion by erwanlpfr** [openlayers/openlayers#14884] [General]
Hello,

I am currently working on having 7+k points with Webgl.

In my map, I have a `LayerGroup -> LayerGroup -> IconLayerWebGL and BadgeLayerWebGL` with these two styles.

To achieve that, I have this style (obstructed a bit for the example).

```typescript
const IconSize = {
  width: 42.5,
  height: 50,
};

const badgeSize = {
  width: 17,
  height: 17,
};

// src are svg files loaded in JS
const rawImgData: Parameters<typeof generateSprites>[0] = [{
  height: IconSize.height * 2,
  width: IconSize.width * 2,
  src: Marker,
}, {
  height: IconSize.height * 2,
  width: IconSize.width * 2,
  src: APIMarker,
}, {
  height: IconSize.height * 2,
  width: IconSize.width * 2,
  src: example3Marker,
}];
const Sprite = await generateSprites(rawImgData);

const labels = ['?', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '9+'];

const sourceIndex: keyof FeatureProperties = 'source';
const statusIndex: keyof FeatureProperties = 'status';
const displayIndex: keyof FeatureProperties = 'display';
const opacityIndex: keyof FeatureProperties = 'opacity';
const badgeIndex: keyof FeatureProperties = 'badge';

/**
 * Can be mutated
 */
export const VariablesStyle = {
  // Brings atomic style settings for each sources
  // Displaying of the point
  example2Display: 1,
  example3Display: 1,
  example1Display: 1,
  // Opacity of the point
  example2Opacity: 1,
  example3Opacity: 1,
  example1Opacity: 1,
} satisfies LiteralStyle['variables'];

export const updateVariablesStyle = (
  style: Partial<typeof VariablesStyle>,
  map?: Map,
) => {
  Object.assign(VariablesStyle, style);
  map?.render();
};

export const BadgeStyle = {
  variables: VariablesStyle,
  filter: ['all',
    ['>', ['get', displayIndex, 'number'], 0],
    ['!=', ['get', statusIndex, 'string'], Status.REJECTED],
    ['match', ['get', sourceIndex, 'string'],
      Source.EXAMPLE_1, ['>', ['var', 'example1Display'], 0],
      Source.EXAMPLE_2, ['>', ['var', 'example2Display'], 0],
      Source.EXAMPLE_3, ['>', ['var', 'example3Display'], 0],
      true,
    ],
  ],
  symbol: {
    symbolType: 'image',
    size: ['interpolate',
      ['exponential', 2.5],
      ['zoom'],
      ZoomLevel.MIN, 1,
      ZoomLevel.MEDIUM, badgeSize.width / 2,
      ZoomLevel.FULL, badgeSize.width,
    ],
    opacity: ['case',
      ['>=', ['get', opacityIndex, 'number'], 0],
      ['interpolate',
        ['exponential', 2.5],
        ['zoom'],
        ZoomLevel.MIN, 0,
        ZoomLevel.MEDIUM, ['get', opacityIndex, 'number'],
        ZoomLevel.FULL, ['get', opacityIndex, 'number'],
      ],
      ['match', ['get', sourceIndex, 'string'],
        Source.EXAMPLE_1, ['var', 'example1Opacity'],
        Source.EXAMPLE_2, ['var', 'example2Opacity'],
        Source.EXAMPLE_3, ['var', 'example3Opacity'],
        1,
      ],
    ],
    rotateWithView: false,
    offset: ['array',
      ['interpolate',
        ['exponential', 2.5],
        ['zoom'],
        ZoomLevel.MIN, 0,
        ZoomLevel.MEDIUM, (IconSize.width * 0.25) / 2,
        ZoomLevel.FULL, IconSize.width / 2,
      ],
      ['interpolate',
        ['exponential', 2.5],
        ['zoom'],
        ZoomLevel.MIN, 0,
        ZoomLevel.MEDIUM, IconSize.width * 0.25,
        ZoomLevel.FULL, IconSize.width,
      ],
    ],
    src: generateBadgeSprites(labels, badgeSize).toDataURL(),
    textureCoord: ['match',
      ['get', badgeIndex, 'string'],
      '?', [0, 0, 1 / 12, 1],
      ...labels.map((label) => {
        if (label === labels[0]) {
          return ['?', [0, 0, 1 / 12, 1]];
        }

        if (labels.includes(label)) {
          const index = labels.indexOf(label);
          const start = index / labels.length;
          const end = (index + 1) / labels.length;

          return [label, [start, 0, end, 1]];
        }

        return [label, [11 / 12, 0, 1, 1]];
      }).flat(),
      [11 / 12, 0, 1, 1],
    ],
  },
} satisfies LiteralStyle;

export const Style = {
  variables: VariablesStyle,
  filter: ['all',
    ['>', ['get', displayIndex, 'number'], 0],
    ['!=', ['get', statusIndex, 'string'], Status.REJECTED],
    ['match', ['get', sourceIndex, 'string'],
      Source.EXAMPLE_1, ['>', ['var', 'example1Display'], 0],
      Source.EXAMPLE_2, ['>', ['var', 'example2Display'], 0],
      Source.EXAMPLE_3, ['>', ['var', 'example3Display'], 0],
      true,
    ],
  ],
  symbol: {
    symbolType: 'image',
    size: ['array',
      ['interpolate',
        ['exponential', 2.5],
        ['zoom'],
        ZoomLevel.MIN, 0,
        ZoomLevel.MEDIUM, IconSize.width * 0.25,
        ZoomLevel.FULL, IconSize.width,
      ],
      ['interpolate',
        ['exponential', 2.5],
        ['zoom'],
        ZoomLevel.MIN, 0,
        ZoomLevel.MEDIUM, IconSize.height * 0.25,
        ZoomLevel.FULL, IconSize.height,
      ],
    ],
    opacity: ['case',
      ['>=', ['get', opacityIndex, 'number'], 0],
      ['interpolate',
        ['exponential', 2.5],
        ['zoom'],
        ZoomLevel.MIN, 0,
        ZoomLevel.MEDIUM, ['get', opacityIndex, 'number'],
        ZoomLevel.FULL, ['get', opacityIndex, 'number'],
      ],
      ['match', ['get', sourceIndex, 'string'],
        Source.EXAMPLE_1, ['var', 'example1Opacity'],
        Source.EXAMPLE_2, ['var', 'example2Opacity'],
        Source.EXAMPLE_3, ['var', 'example3Opacity'],
        1,
      ],
    ],
    rotateWithView: false,
    offset: ['array',
      0,
      ['interpolate',
        ['exponential', 2.5],
        ['zoom'],
        ZoomLevel.MIN, 0,
        ZoomLevel.MEDIUM, (IconSize.width * 0.25) / 2,
        ZoomLevel.FULL, IconSize.width / 2,
      ],
    ],
    src: Sprite.toDataURL(),
    textureCoord: ['match', ['get', sourceIndex, 'string'],
      Source.EXAMPLE_1, [0, 0, 1 / 3, 1],
      Source.EXAMPLE_2, [1 / 3, 0, 2 / 3, 1],
      Source.EXAMPLE_3, ['match',
        ['get', statusIndex, 'string'],
        Status.APPROVED, [0, 0, 1 / 3, 1],
        [2 / 3, 0, 1, 1],
      ],
      [0, 0, 1 / 3, 1],
    ],
  },
} satisfies LiteralStyle;
```

In this situation, in a 60fps screen, I can have a stable 58-60fps. Which is excellent in addition of vectorial layers.
In another hand, I have a 120 fps screen with here, I see a drop from 120 to 70 when moving the map.

So, here, as discussion, I would to know if you ever had similar issues and had ideas to improve this.
On my side, I read about suggestions from Mapbox literal expressions which is good but maybe not accurate in this situation.

 Let me know if you have any tips!
Thank you!!

**Comment by jahow** [comment]
Hi! both your styles are quite complex but it's hard to tell what could help performance without knowing more (data, code...). Would you be able to make a running example that shows this performance drop? also note that 60fps is really what we're targeting with WebGL renderers, 120fps would be much harder to achieve :slightly_smiling_face: 

One thing you could do is use the `WebGLVectorLayerRenderer` (if you're not already), it is generally more performance and will probably replace `WebGLPointsLayerRenderer` eventually. Also this renderer can take an array of styles, so you could have three styles for each `Source.EXAMPLE_X` value instead of repeating the same `match` operator many times.

We still have work planned for improving performance, including filtering features ahead of the rendering pipeline which should offer a lot of gain.

**Reply by erwanlpfr** [reply]
Thank you for your reply !
I began by migrating to this `WebGLVectorLayerRenderer`. But actually, I get nothing on the layer.
It seems that all the instruction of my style are not listened. If I comment opacity and filter, I get black squares and ignoring my texture.

Filter and opacity set at just 1 also make the features not appearing.

Do you have an example with a sprite?
I try to mix these two examples without success and same problem : 

https://codesandbox.io/s/webgl-vector-layer-forked-dtrtpq?file=/main.js:338-411
https://codesandbox.io/s/icon-sprite-webgl-forked-9yxvkn?file=/main.js

Thank you for your answer, it helps a lot !

**Reply by jahow** [reply]
Oh yes sorry, indeed icon rendering does not work yet with the WebGLVectorLayerRenderer; this is fixed in https://github.com/openlayers/openlayers/pull/14883

**Reply by erwanlpfr** [reply]
Oh right ! If I understand the PR, the `symbol` attribute will break and the flat style will be the now common way?
Like, the `src` attribute would be dynamic?

Looks superb, you are achieving a really good work !
If I have time, I would like to help to migrate/give new examples.

EDIT : I did a migration from your PR to test.
Right now, I don't get any feedback while using `map.getFeaturesAtPixel` or `map.forEachFeatureAtPixel`.
Is there a prop to set somewhere to get the hit detection or this layer don't receive them?

**Reply by jahow** [reply]
This renderer also does not have hit detection enabled yet unfortunately.

> Oh right ! If I understand the PR, the symbol attribute will break and the flat style will be the now common way?
Like, the src attribute would be dynamic?

`icon-src` will not support expressions initially, that might come later but right now the preferred to have dynamic icons is to use a sprite sheet. But other than that yes, we're moving all symbols to the common "flat" format!

I'd be very interested in hearing if you notice anything in terms of performance with the PR. Also, don't hesitate to mention anything you notice in the PR, that will help a lot! thanks :)

**Reply by erwanlpfr** [reply]
> This renderer also does not have hit detection enabled yet unfortunately.

Okay, understood ! I have to say that It would be a really great feature.

> icon-src will not support expressions initially, that might come later but right now the preferred to have dynamic icons is to use a sprite sheet. But other than that yes, we're moving all symbols to the common "flat" format!

A huge improvement !

> I'd be very interested in hearing if you notice anything in terms of performance with the PR. Also, don't hesitate to mention anything you notice in the PR, that will help a lot! thanks :)

Okay, for now, the migrations of my styles weren't painful at all. Even more, the styles are less verbose and makes easier the maintenance.
I am using an M1 with Brave/Safari/Chrome/Firefox.
I am really surprised since it reaches the 110–120 fps constantly 👍🏻 !

The only thing I needed time to guess the differences with : `icon-width`, `icon-height`, `icon-size`, `icon-img-size`.
And right now, I would not say that I got all of them. 

After that, I really appreciate this feature !
